pkg/security/resolvers/hash/resolver.go

```diff
@@ -0,0 +1,361 @@
+		return nil
+	}
+
+	// check state
+	if file.HashState == model.Done {
+		return file.Hashes
+	}
+	if file.HashState != model.NoHash && file.HashState != model.HashWasRateLimited {
+		// this file was already processed and an error occurred, nothing else to do
+		return nil
+	}
+
+	// check if the resolver is allowed to hash this event type
+	if !slices.Contains[model.EventType](resolver.opts.EventTypes, eventType) {
+		file.HashState = model.EventTypeNotConfigured
+		resolver.hashMiss[eventType][model.EventTypeNotConfigured].Inc()
+		return nil
+	}
+
+	if err := resolver.hash(eventType, process, file); err != nil {
+		resolver.hashMiss[eventType][model.UnknownHashError].Inc()
+		seclog.Errorf("hash computation failed: %v", err)
+		file.HashState = model.UnknownHashError
+		return nil
+	}
+
+	return file.Hashes
+}
+
+// getHashFunction returns the hash function for the provided algorithm
+func (resolver *Resolver) getHashFunction(algorithm model.HashAlgorithm) hash.Hash {
+	switch algorithm {
+	case model.SHA1:
+		return sha1.New()
+	case model.SHA256:
+		return sha256.New()
+	case model.MD5:
+		return md5.New()
+	default:
+		return nil
+	}
+}
+
+// hash hashes the provided file event
+func (resolver *Resolver) hash(eventType model.EventType, process *model.Process, file *model.FileEvent) error {
+	if !file.IsPathnameStrResolved || len(file.PathnameStr) == 0 {
+		resolver.hashMiss[eventType][model.PathnameResolutionError].Inc()
+		return fmt.Errorf("pathname resolution error: can't hash a FileEvent that hasn't been properly resolved")
+	}
+
+	// check if the hash(es) of this file is in cache
+	if resolver.cache != nil {
+		cacheEntry, ok := resolver.cache.Get(file.PathKey)
+		if ok {
+			file.HashState = cacheEntry.state
+			file.Hashes = cacheEntry.hashes
+			resolver.hashCacheHit[eventType].Inc()
+			return nil
+		}
+	}
+
+	rootPIDs := []uint32{process.Pid}
+	if resolver.cgroupResolver != nil {
+		w, ok := resolver.cgroupResolver.GetWorkload(process.ContainerID)
+		if ok {
+			rootPIDs = w.GetPIDs()
+		}
+	}
+
+	// open the target file
+	var lastErr error
+	var f *os.File
+	for _, pidCandidate := range rootPIDs {
+		f, lastErr = os.Open(filepath.Join(utils.ProcRootPath(int32(pidCandidate)), file.PathnameStr))
+		if lastErr == nil {
+			break
+		}
+	}
+	if lastErr != nil {
+		if os.IsNotExist(lastErr) {
+			file.HashState = model.FileNotFound
+			resolver.hashMiss[eventType][model.FileNotFound].Inc()
+			return nil
+		}
+		return fmt.Errorf("couldn't open file: %w", lastErr)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return fmt.Errorf("couldn't stat file: %w", err)
+	}
+
+	// is this a regular file ?
+	if !fi.Mode().IsRegular() {
+		file.HashState = model.Done
+		return nil
+	}
+
+	// check file size
+	if fi.Size() > resolver.opts.MaxFileSize {
+		resolver.hashMiss[eventType][model.FileTooBig].Inc()
+		file.HashState = model.FileTooBig
+		return nil
+	}
+
+	// check the rate limiter
+	if !resolver.limiter.Allow() {
```

Comment:
Discussed offline: this rate limiter is meant to count the number of hashes that we computed, if we move this check up, it will count the number of times we checked if a file should be hashed.